pylabrobot/resources/functional.py

```diff
@@ -0,0 +1,59 @@
+import os
+import json
+import random
+from collections import deque
+from typing import List, Optional
+from pylabrobot.resources.tip_rack import TipRack
+
+
+async def linear_tip_spot_generator(tip_racks: List[TipRack],
+                                    cache_file_path: Optional[str]=None):
+  """ Tip spot generator with disk caching. Linearly iterate through all tip spots and
+  raise StopIteration when all spots have been used. """
+  tip_rack_idx, tip_spot_idx = 0, 0
+  if cache_file_path is not None and os.path.exists(cache_file_path):
+    with open(cache_file_path, "r", encoding="utf-8") as f:
+      data = json.load(f)
+      tip_rack_idx, tip_spot_idx = data["tri"], data["tsi"]
+      print("loaded tip idx from disk:", data)
+
+  while tip_rack_idx < len(tip_racks):
+    while tip_spot_idx < 96:
+      tip_spot_idx += 1
+      if cache_file_path is not None:
+        with open(cache_file_path, "w", encoding="utf-8") as f:
+          json.dump({"tri": tip_rack_idx, "tsi": tip_spot_idx}, f)
+      yield tip_racks[tip_rack_idx][tip_spot_idx]
+
+    tip_rack_idx += 1
+    tip_spot_idx = 0
+
+
+async def randomized_tip_spot_generator(tip_racks: List[TipRack], K: int,
+                                        cache_file_path: Optional[str]=None):
+  """ Randomized tip spot generator with disk caching. Don't return tip spots that have been
+  sampled in the last K samples. """
+  
+  all_tip_spots = [rack.get_item(spot_idx) for rack in tip_racks for spot_idx in range(rack.num_items)]
+  recently_sampled = deque(maxlen=K)
+
+  if cache_file_path is not None and os.path.exists(cache_file_path):
+    with open(cache_file_path, "r", encoding="utf-8") as f:
+      data = json.load(f)
```

Comment:
have you considered converting this into a dataclass? this allows readers of the code to see what was intended and validates input

Reply:
what would the dataclass be in this case? it's a single parameter

Reply:
for storing the data on disk?

Reply:
for bringing it back from disk.

this ensure the json wasn't corrupted and allows people to see what it looks like. 

It's a nice to have, not a must have.

Reply:
probably in a future version